tests/unit/test_symmetric_crypto.cpp

```diff
@@ -0,0 +1,312 @@
+    }
+
+    encryptor->update(plaintext);
+    auto ciphertext = encryptor->finish();
+    auto iv = encryptor->getIV();
+
+    std::vector<uint8_t> tag;
+    auto authenticatedEncryptor = dynamic_cast<AuthenticatedEncryptionI*>(encryptor.get());
+    if (authenticatedEncryptor) {
+        tag = authenticatedEncryptor->getAuthTag();
+    }
+
+    auto decryptorBuilder = AESCipherBuilder{operationMode, _secretKey}.setIV(iv);
+    std::shared_ptr<SymmetricCipher> decryptor;
+    if (isAuthenticatedCipherMode(operationMode)) {
+        decryptor = encryptoBuilder.buildAuthenticatedDecryptor();
+    }
+    else {
+        decryptor = encryptoBuilder.buildDecryptor();
+    }
+
+    auto authenticatedDecryptor = dynamic_cast<AuthenticatedEncryptionI*>(decryptor.get());
+    if (authenticatedDecryptor) {
+        authenticatedDecryptor->setAuthTag(tag);
+    }
+
+    decryptor->update(ciphertext);
+    auto decryptedText = decryptor->finish();
+
+    // Then: Decrypted text should match the plaintext
+    ASSERT_THAT(decryptedText, ::testing::ElementsAreArray(plaintext));
+}
+
+
+static std::vector<EncrytDecryptTestData> prepareTestDataForMode(SymmetricCipherMode mode)
+{
+    static const std::vector<std::string> PlaintextStringsOfDifferentLength
+            {"0",
+             "0123456789", // less than a block
+             "0123456789123456", // a block
+             "01234567891234567890" // more than a block
+            };
+
+    std::vector<EncrytDecryptTestData> testData;
+    for (const auto &elem : PlaintextStringsOfDifferentLength) {
+        testData.emplace_back(std::make_tuple(mode, elem));
+    }
+    return testData;
+}
+
+INSTANTIATE_TEST_SUITE_P(GCM,
+                         SymmetricCipherTest,
+                         testing::ValuesIn(prepareTestDataForMode(SymmetricCipherMode::GCM)));
+
+INSTANTIATE_TEST_SUITE_P(CBC,
+                         SymmetricCipherTest,
+                         testing::ValuesIn(prepareTestDataForMode(SymmetricCipherMode::CBC)));
+
+INSTANTIATE_TEST_SUITE_P(CTR,
+                         SymmetricCipherTest,
+                         testing::ValuesIn(prepareTestDataForMode(SymmetricCipherMode::CTR)));
+
+
+TEST_P(SymmetricCipherAdvancedTest, encryptMultipleChunksAndDecrypt)
+{
+    auto operationMode = GetParam();
+
+    // Given: long plaintext string
+
+    // When: Read and encrypt plaintext in chunks. Then decrypt ciphertext.
+    auto encryptoBuilder = AESCipherBuilder{operationMode, _secretKey};
+    std::shared_ptr<SymmetricCipher> encryptor;
+    if (isAuthenticatedCipherMode(operationMode)) {
+        encryptor = encryptoBuilder.buildAuthenticatedEncryptor();
+    }
+    else {
+        encryptor = encryptoBuilder.buildEncryptor();
+    }
+
+    size_t packetSize = 1024 * 4;
+    auto packetIterator = _plaintext.begin();
+    while (packetIterator < _plaintext.end() - packetSize) {
+        encryptor->update({packetIterator, packetIterator + packetSize});
```

Comment:
not nice to read. prefix with std::vector?

Reply:
why more code would be nicer to read?